Test for CountryZoneMember and StateZoneMember (model/schema)

Jeshan7 opened the pull request:
## Why?
  - Test cases were less for country and state zone member(model/schema)
  - Coverage was less for country and state zone member(model/schema)

## This change addresses the need by:
 - Wrote new test cases for update_changset in the schema for both country zone
   and state zone member.(Schema)
 - Wrote test cases for get/1, get_all/0, member_changesets/2,
   remove_members_query/2, common_zone_query/2 for both country and
   state zone member.(Model)

[delivers #163514116]

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read [CONTRIBUTING.md][contributing].
- [ ] My code follows the [style guidelines][contributing] of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [ ] I have updated the documentation wherever necessary.
- [ ] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

Jeshan7 committed:
[#163514116]

Prior to this change,
  - Test cases were less for country and state zone member(model/schema)
  - Coverage was less for country and state zone member(model/schema)

This change
 - Wrote new test cases for update_changset in schema for both country zone
   and state zone member.(Schema)
 - Wrote test cases for get/1, get_all/0, member_changesets/2,
   remove_members_query/2, common_zone_query/2 for both country and
   state zone member.(Model)

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
   end
 
+  describe "get/1" do
+    test "zone with id", %{zone: zone} do
```
Add the failure one too.

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
+  end
+
+  describe "get_all/0" do
+    @tag state_zone_count: 3,
```
what's the point of setting `state_zone_count`

Jeshan7 replied:
To ensure that in the list of structs of country_zone_member and state_zone_member, the get_all/0 function retrieves only those structs which have zone_type set as "C".Similarly in the state_zone_member model also.

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
+    test "zones  successfully", %{zones: zone} do
+      country_zones = CountryZone.get_all()
+      cz = Enum.drop_while(country_zones, fn x -> x.zone_type == "C" end)
+      assert cz == []
```
you could have easily checked the number of countries set. Please also add test when none are set

Jeshan7 replied:
In this test, I have used zones function present in the factory instead of country_zone function below. get_all/0 retrieves all the zones so while checking I have to assert the count with one extra count added to the country_zone_count tag.  
assert sz(count obtained) == 3 where country_zone_count is set to 2.
I don't think so then it would be an appropriate test because we are changing variables according to our need in the test.

arjun289 replied:
`CountryZone.get_all()` returns a list of country zones so you just need to check all the members of this list have zone type as "C". You could have used `Enum.all?(all, fn one -> one.zone_type == "C" end)`

Jeshan7 replied:
ok

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
+  end
+
+  describe "member_changesets/2" do
+    test "successfully returns a CountryZoneMember changeset", %{zone: zone, countries: countries} do
```
I think the statement should be `returns a valid changeset`

arjun289 replied:
Also, add scenario when it would be invalid.

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
+
+  describe "remove_members_query/2" do
+    @tag country_zone_count: 3
+    test "returns a query successfully", %{zone: zone, countries: countries} do
```
statement should be `returns a valid query`

arjun289 commented on apps/snitch_core/test/data/model/zone/state_zone_test.exs:
```diff
@@ -84,6 +88,69 @@ defmodule Snitch.Data.Model.StateZoneTest do
+         country_zone_count: 3
+    test "zones  successfully", %{zones: zone} do
+      state_zones = StateZone.get_all()
+      sz = Enum.drop_while(state_zones, fn x -> x.zone_type == "S" end)
```
Any particular reason you are handling it this way, I am not able to see the purpose.

Jeshan7 replied:
I have used it to access all the struct records inside the list(state_zones). I have used it because only the zone_type field was common to all the records present inside the list(state_zones).

arjun289 commented on apps/snitch_core/test/data/schema/zone/country_zone_member_test.exs:
```diff
@@ -45,4 +60,39 @@ defmodule Snitch.Data.Schema.CountryZoneMemberTest do
+      ]
+    end
+
+    test "successfully for valid params", %{cs: cs} do
```
should be `returns a valid changeset` instead the statement

arjun289 reviewed (commented):
I would suggest naming the tests properly. The "describe" along with the test should return a proper statement. Also, consider not using "describe" for functions which have only one test case most likely the `get_all` ones

Jeshan7 committed:
[#163514116]

Prior to this change,
  In CountryZone and StateZone Model
   - Few test statements were wrong.
   - For get/1, member_changeset/2 and get_all/0 failure test were not present.
   - In get_all/0 test, i was dropping all the records of the list.

  In CountryZone and StateZone Schema
   - One test staement was needed to be changed(in update_changeste/2)
This change
  In CountryZone and StateZone Model
   - Corrected the test statements.
   - For get/1, member_changeset/2 and get_all/0 failure tests are now present.
   - In get_all/0 test now it checks for total number of reqired records.

  In CountryZone and StateZone Schema
   - One test staement was changed(in update_changeste/2) to returns
     a valid changeset

Jeshan7 committed:
[#163514116]

Prior to this change,
 CountryZone and StateZone(Model)
  - In get/1, zone was deleted first and the asserted.

This change
 CountryZone and StateZone(Model)
  - Now in get/1, -1 is passed and after that this function is asserted.

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -77,6 +81,85 @@ defmodule Snitch.Data.Model.CountryZoneTest do
+
+    test "fails for invalid id", %{zone: zone} do
+      {:ok, _} = CountryZone.delete(zone)
+      assert CountryZone.get(zone.id) == {:error, :zone_not_found}
```
Instead of deleting the zone you can just pass `-1` as id and it will fail

arjun289 commented on apps/snitch_core/test/data/model/zone/state_zone_test.exs:
```diff
@@ -84,6 +88,85 @@ defmodule Snitch.Data.Model.StateZoneTest do
+    end
+
+    test "fails for invalid id", %{zone: zone} do
+      {:ok, _} = StateZone.delete(zone)
```
see the comment on country `get`

Jeshan7 committed:
[#163514116]

Prior to this change,
  - Import and alias, both were used for Ecto.Query .

This change
  - Now only alias is used for Ecto.Query .

arjun289 commented on apps/snitch_core/test/data/model/zone/country_zone_test.exs:
```diff
@@ -3,11 +3,15 @@ defmodule Snitch.Data.Model.CountryZoneTest do
 
   import Snitch.Factory
+  import Ecto.Query
+  alias Ecto.Query
```
why both `import` and `alias`? Use any one of them.
